### ISSUE TITLE
Add sequence_nr to aot_autograd to enable mapping of backward ops in inductor kernels to their corresponding forward op 

### ISSUE THREAD

**Issue by alexmsettle** [pytorch/pytorch#102375]
### 🚀 The feature, motivation and pitch

Eager execution currently has a sequence id (sequence_nr) which is populated by autograd.  The sequence id is a unique id which maps a backward op to its corresponding forward op.  This is useful for doing performance analysis.  Inductor uses aot_autograd to generate the backward ops.  Even though aot_autograd calls into autograd it looks like the entry point is different,  **torch.autograd.grad()** is the entry point for aot_autograd which happens after the sequence_id setup code is normally executed in autograd.  The variable in autograd is called sequence_nr.

My proposal is to re-introduce the sequence id to the aot autograd path to help bring performance analysis capabilities for inductor up to parity with eager mode. 

### Alternatives

It might be possible to use string parsing of the stack trace in fx/traceback.py but I'm not sure if this would provide sufficient information to accurately do the mapping.

### Additional context

I implemented aot sequence id already in my local copy of pytorch.  This is part of a larger effort to improve profiler visibility into inductor kernels.  Here's a screen shot below that shows an example of how the seq_id is being used.  I inserted nvtx markers into the generated inductor/triton kernels.  You can see in the pictures below that ***aot_seq_id 4 and 5*** represent specific batch norm and relu ops that were fused into a triton kernel in the forward pass.  In the bwd pass you can see that ***aot_seq_id 4 and 5*** represent the corresponding backward ops from the same bn and relu as in the forward pass.

![inductor_fwd_pass](https://github.com/pytorch/pytorch/assets/37422826/fb58332c-c92f-4234-85d1-78d9daecdf4c)

![inductor_bwd_pass](https://github.com/pytorch/pytorch/assets/37422826/59c5f54b-486b-4869-82e0-fa7f7317cd96)

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @robieta @chaekit @aaronenyeshi @ngimel @nbcsm @guotuofeng @guyang3532 @gaoteng-git @tiffzhaofb @dzhulgakov @davidberard98 @msaroufim @wconstab @bdhirsh @anijain2305

**Comment by ezyang** [comment]
Sure, send it in.

**Comment by ngimel** [comment]
cc @davidberard98

**Comment by davidberard98** [comment]
@alexmsettle , this looks great! We're also interested in displaying this sort of information in the torch profiler, so I'd be interested to see how you are adding the nvtx markers. If you have a code pointer for where you're adding them, that would be great (or if you have a PR coming, that's good too).

**Comment by alexmsettle** [comment]
Thanks! I’m on vacation right now. I’ll submit a PR next week when I get
back.

**Comment by alexmsettle** [comment]
> @alexmsettle , this looks great! We're also interested in displaying this sort of information in the torch profiler, so I'd be interested to see how you are adding the nvtx markers. If you have a code pointer for where you're adding them, that would be great (or if you have a PR coming, that's good too).

Hi @davidberard98 I started a PR for the sequence ID tracking in aot but it is currently separate from the profiling logic I added to the triton kernels.  Should I start a separate issue for the profiling or can I start a draft PR?

**Comment by davidberard98** [comment]
@alexmsettle awesome! You are free to open a draft PR or an issue, whichever you prefer

**Comment by alexmsettle** [comment]
> @alexmsettle awesome! You are free to open a draft PR or an issue, whichever you prefer

Hi @davidberard98 - I just filed this issue #103552 to cover the nvtx markers which get inserted in triton kernels
